spring-xd-dirt/src/main/java/org/springframework/xd/dirt/cluster/ContainerFilter.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.xd.dirt.cluster;
+
+import java.util.List;
+
+import org.springframework.xd.module.ModuleDescriptor;
+
+/**
+ * A filter interface that applies filter for the given containers and returns
+ * eligible containers for deployment.
+ *
+ * @author Ilayaperumal Gopinathan
+ */
+public interface ContainerFilter {
```

Comment:
I started down this path of refactoring the ContainerFilter to just be another ContainerMatcher (and I have a CompositeContainerMatcher that delegates), but I'm actually questioning that approach now. The main concern is that DefaultContainerMatcher is very particular about the way results are ordered and finalized. It's important for the round-robin distribution of module deployments. If we are providing an extension point, we do not want to lose that control. Therefore, the approach in this PR does seem correct in the sense that the *filtering* of containers happens within the scope of the DefaultContainerMatcher itself, after the initial criteria check but *before* the final distributeForRequestedCount(). So rather than having a delegating ContainerMatcher that iterates a list, we really do want this filtering to happen in that well-defined scope of the existing algorithm. Given that, I'm actually much less opposed to the idea of defining an explicit (even if similar) interface that is intended as a "safe" extension point within those controlled bounds. Thoughts?